### PR TITLE
Stop rct overwrite

### LINKED ISSUE #81: Don't overwrite RCT assignment

### LINKED ISSUE #87: Create fix quartile bins for each poli

### LINKED ISSUE #88: create script to create fixed bins and train model

### PR DESCRIPTION
**Pull request by WoltersEric**
closes #81 
closes #87 
closes #88

### CODE REVIEW

**Review by sjoerd-de-vries — COMMENTED** [review]
Two comments for app_helpers.py:
- l126 sorting values and then l129 drop values but keep first results (I believe) in keeping the bottom prediction, not the top
- l135 assigning treatment arms in this way causes the 0 arm to be assigned an additional person 50% of the time. might cause some imbalance over time.

File: src/noshow/api/app_helpers.py [hunk 1]
```diff
@@ -99,25 +100,54 @@ def create_treatment_groups(
...
     # only keep top prediction per patient for treatment/control split
     deduplicated = predictions.drop_duplicates(subset="pseudo_id", keep="first")
+    # only keep patients without treatment group
+    deduplicated = deduplicated[deduplicated["treatment_group"].isnull()]
 
     # Create stratified randomization in control and treatment groups
     deduplicated = deduplicated.sort_values(["hoofdagenda", "prediction"])
     deduplicated["treatment_group"] = deduplicated.groupby(
         ["hoofdagenda", "score_bin"], observed=True
     )["prediction"].transform(lambda x: np.arange(len(x)) % 2)
```
Comment (LauraVeerhoek) [inline src/noshow/api/app_helpers.py:137]: Suggestie nadat we het besproken hebben: maak hier een random sample van.

File: src/noshow/api/app_helpers.py [hunk 2]
```diff
@@ -99,25 +100,54 @@ def create_treatment_groups(
...
 
-    predictions = pd.merge(
-        predictions,
+    # update predictions with treatment groups from deduplicated based on pseudo_id
+    merged = predictions.merge(
```
Comment (LauraVeerhoek) [inline src/noshow/api/app_helpers.py:140]: Bij het consistent houden van de treatment group door een patiënt er standaard in 1 te houden, zou ik controleren hoe veel invloed dit heeft op het formaat/de verhouding van je control en treatment group. Misschien kun je retrospectief met wat data van afgelopen maand kijken.

File: tests/test_api.py [hunk 3]
```diff
@@ -42,6 +42,11 @@ async def test_predict_endpoint(monkeypatch):
...
     monkeypatch.setattr(app, "load_model", fake_model)
     monkeypatch.setattr(app, "delete", lambda x: x)
+    # patch create treatment groups and add column to the dataframe
+    monkeypatch.setattr(
+        app, "create_treatment_groups", lambda x, y: x.assign(treatment_group=1)
+    )
+
     output = await predict(appointments_json, "2023-01-05", FakeDB())
     output_df = pd.DataFrame(output)
-    assert output_df.shape == (4, 16)
```
Comment (LauraVeerhoek) [inline tests/test_api.py:47]: Wat was de reden dat je dit veranderd hebt?

Comment (LauraVeerhoek) [reply tests/test_api.py:47]: Besproken in persoon.

**Review by LauraVeerhoek — APPROVED** [review]

**Comment by WoltersEric** [comment]
added fixed bins implementation closes #87

File: pyproject.toml [hunk 4]
```diff
@@ -4,7 +4,7 @@ build-backend = "setuptools.build_meta"
 
 [project]
 name = "noshow"
-version = "1.1.0"
+version = "1.1.1"
 authors = [
   { name="Ruben Peters", email="r.peters-7@umcutrecht.nl" }
```
Comment (sjoerd-de-vries) [inline pyproject.toml:9]: Volgens mij mag jij er ook wel bij hoor Eric

File: .gitignore [hunk 5]
```diff
@@ -162,3 +162,6 @@ dmypy.json
 
 # Cython debug symbols
 cython_debug/
+
+# fixed bins
+data/processed/*.json
```
Comment (sjoerd-de-vries) [inline .gitignore:167]: Wil je de bins niet op git hebben staan? Lijkt me op zich juist wel een goed idee.

File: src/noshow/api/app_helpers.py [hunk 6]
```diff
@@ -119,35 +134,31 @@ def create_treatment_groups(
...
+        predictions[mask]
+        .groupby(["hoofdagenda", "score_bin"])["prediction"]
+        .transform(lambda x: (np.arange(len(x)) + random.randint(0, 1)) % 2)
     )
 
-    # Update the treatment_group values where there's a new value
-    predictions["treatment_group"] = merged["treatment_group_new"].combine_first(
-        predictions["treatment_group"]
+    # For every new patient assign the treatment group based on the mode
+    predictions.loc[mask, "treatment_group"] = (
```
Comment (sjoerd-de-vries) [inline src/noshow/api/app_helpers.py:156]: Als ik het goed begrijp gebeurt dit zodat als een patiënt voor het eerst een voorspelling krijgt en er meerdere afspraken op die dag zijn, de patiënt toch maar 1 groep geassigned krijgt? Misschien is het netter om dit voor de vorige stap af te handelen door alleen de eerste afspraken te assignen oid?

File: notebooks/train_model.ipynb [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    "featuretable[\"no_show\"] = (\n",
+    "    featuretable[\"no_show\"].replace({\"no_show\": \"1\", \"show\": \"0\"}).astype(int)\n",
+    ")\n",
+    "featuretable[\"hour\"] = featuretable[\"hour\"].astype(\"category\")\n",
+    "featuretable[\"weekday\"] = featuretable[\"weekday\"].astype(\"category\")\n",
+    "\n",
+    "print(featuretable.dtypes)\n",
+    "X, y = featuretable.drop(columns=\"no_show\"), featuretable[\"no_show\"]\n",
+    "X_train, X_test, y_train, y_test = train_test_split(\n",
+    "    X, y, test_size=0.2, random_state=0, shuffle=False\n",
```
Comment (sjoerd-de-vries) [inline notebooks/train_model.ipynb:100]: Doe je shuffle = False omdat je een validatie over de tijd wil?

File: notebooks/model_experiments.ipynb [hunk 8]
```diff
@@ -207,6 +211,307 @@
...
+    "plt.ylabel(\"Count\")\n",
+    "plt.title(\"Distribution of Treatment Groups per hoofdagenda and score_bin for first appointment\")\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Evaluate the stratification and visualize new distributions"
```
Comment (sjoerd-de-vries) [inline notebooks/model_experiments.ipynb:442]: nog even samen naar de plots kijken?

**Review by sjoerd-de-vries — APPROVED** [review]
Looks good, some minor comments

File: notebooks/train_model.ipynb (hunk 7 above)
Comment (WoltersEric) [reply notebooks/train_model.ipynb:100]: Dit had True moeten zijn

File: .gitignore (hunk 5 above)
Comment (WoltersEric) [reply .gitignore:167]: ik dacht dit via dvc te doen

File: src/noshow/api/app_helpers.py (hunk 6 above)
Comment (WoltersEric) [reply src/noshow/api/app_helpers.py:156]: besproken

**Review by LauraVeerhoek — APPROVED** [review]
